### DISCUSSION TITLE
tokio::io::copy slower than std io::copy

### DISCUSSION THREAD

**Discussion by radumarias** [tokio-rs/tokio#6553] [Q&A]
I did a file copy speed test between these 2 and seems `tokio::io::copy` is much slower than sync std `io::copy`

I noticed this while implementing my own tokio `AsyncWrite`.

Here are the results on my machine for a file of `894MB`
```
filesize = 894MB
tokio write duration = 2.661018252s, speed MB/s 336.2394975276158
std write duration = 418.723487ms, speed MB/s 2136.826492286769
```

here is the code I used [playground](https://play.rust-lang.org/?version=stable&mode=release&edition=2021&gist=44db4d88a3aabea2dce237576659746c)

I then tried `tokio-uring`, Indeed better but still not close to std
Tested writing 840 MB of content to a file
```
tokio write duration = 710.079182ms, speed MB/s 1260.0558679162832
```

I understand std is using `copy_file_range` syscall but why is it such a big difference and can't it be improved?

**Comment by Darksonn** [comment]
I explain this on [the URLO thread](https://users.rust-lang.org/t/tokio-copy-slower-than-std-io-copy/111242) that you also posted.
